torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -184,22 +192,42 @@ def __add__(self, other_datapipe):
     @staticmethod
     def _cache_check_fn(data, filepath_fn, hash_dict, hash_type, extra_check_fn):
         filepaths = data if filepath_fn is None else filepath_fn(data)
+        result = True
         if not isinstance(filepaths, (list, tuple)):
             filepaths = [
                 filepaths,
             ]
 
         for filepath in filepaths:
+            cached_file_exists = True
             if not os.path.exists(filepath):
-                return False
-
-            if hash_dict is not None and not _hash_check(filepath, hash_dict, hash_type):
-                return False
-
-            if extra_check_fn is not None and not extra_check_fn(filepath):
-                return False
-
-        return True
+                cached_file_exists = False
+            elif hash_dict is not None and not _hash_check(filepath, hash_dict, hash_type):
+                cached_file_exists = False
+            elif extra_check_fn is not None and not extra_check_fn(filepath):
+                cached_file_exists = False
+
+            if not cached_file_exists:
+                promise_filepath = filepath + ".promise"
+                dirname = os.path.dirname(promise_filepath)
+                if not os.path.exists(dirname):
+                    os.makedirs(dirname)
+
+                with portalocker.Lock(promise_filepath, "a+", flags=portalocker.LockFlags.EXCLUSIVE) as promise_fh:
+                    promise_fh.seek(0)
+                    data = promise_fh.read()
+                    # TODO(VitalyFedyunin): Potentially there is old .promise file from previous failed run, we
+                    # need to somehow propagate uniq session id for dataloader, save and compare it here,
+                    # raising error
+                    file_exists = len(data) > 0
+                    if not file_exists:
+                        result = False
+                        promise_fh.seek(0)
+                        promise_fh.write("[dataloader session uid]")
+                        promise_fh.truncate()
+                        promise_fh.flush()
+
+        return result
```

Comment:
Unfortunately data could be url or something else, it is hard to lock on it.

But this situation is covered. 

Imagine data generates two file namesL file1 file2.

Initial pass (empty FS) will add two locks file1.promise and file2.promise and will go 'False' route.

Now second (and every next) pass will see that files are missing, but will fail to create promise and go into the 'file exists' route, which will led them to the situation when they are waiting for file1.promise and file2.promise to disappear.

Reply:
If it is an URL, is it possible to create and lock `root/URL.promise` in the file system?

I think we should have a similar lock for `HttpReader` to prevent multiple processes from downloading the same file?

Reply:
Oh, I see. `file_exists` flag is used for processes to recognize this file or parent archives are going to be processed by another process.